### PR TITLE
romaan_to_int

### PR DESCRIPTION
**Pull request by Mushtary**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/roman_to_int.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+def roman_to_int(s: str) -> int:
```
Comment (asagulenko) [inline solutions/roman_to_int.py:1]: Please add the intro docstring like this:

#!/usr/local/bin/python3
# -*- coding: utf-8 -*-
"""
Created on 01.01.2025
@author: Anna Shumylina
"""

File: solutions/roman_to_int.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    >>> roman_to_int("IV")
+    4
+    >>> roman_to_int("IX")
+    9
+    >>> roman_to_int("LVIII")
+    58
+    >>> roman_to_int("MCMXCIV")
+    1994
+    """
+
```
Comment (asagulenko) [inline solutions/roman_to_int.py:27]: It's great to have so many doctests, but maybe 3 are enough?

File: solutions/tests/test_roman_to_int.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+"""
+Test suite for the roman_to_int function.
```
Comment (asagulenko) [inline solutions/tests/test_roman_to_int.py:2]: also please add into here

```
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-
"""
Created on 01.01.2025
@author: Anna Shumylina
"""
```

File: solutions/tests/test_roman_to_int.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    def test_roman_to_int_invalid_IIII(self):
+        """Test invalid Roman numeral 'IIII'."""
+        with self.assertRaises(ValueError):
+            roman_to_int("IIII")
+
+    def test_roman_to_int_invalid_ABC(self):
+        """Test invalid non-Roman input 'ABC'."""
+        with self.assertRaises(ValueError):
+            roman_to_int("ABC")
+
```
Comment (asagulenko) [inline solutions/tests/test_roman_to_int.py:54]: Please add comments to each test case

File: solutions/tests/test_roman_to_int.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+Test suite for the roman_to_int function.
+
+This module provides a series of unit tests for the roman_to_int function
+to ensure it correctly converts Roman numerals to integers and handles
+invalid inputs appropriately.
+"""
+
+import sys
+
+sys.path.insert(0, "/Users/mushtary/ET6-foundations-group-03/solutions")
```
Comment (asagulenko) [inline solutions/tests/test_roman_to_int.py:11]: Will this file path work for everyone using the test?

File: solutions/roman_to_int.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    3
+    >>> roman_to_int("IV")
+    4
+    >>> roman_to_int("IX")
+    9
+    >>> roman_to_int("LVIII")
+    58
+    >>> roman_to_int("MCMXCIV")
+    1994
+    """
```
Comment (asagulenko) [inline solutions/roman_to_int.py:26]: It's great to have so many doctests, but maybe 3 are enough?

File: solutions/roman_to_int.py (hunk 1 above)
Comment (asagulenko) [inline solutions/roman_to_int.py:1]: Please add the intro docstring like this before the function definition :

```
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-
"""
Created on 01.01.2025
@author: Anna Shumylina
"""
```

File: solutions/tests/test_roman_to_int.py (hunk 4 above)
Comment (asagulenko) [inline solutions/tests/test_roman_to_int.py:54]: Please add comments to each test case

**Review by asagulenko — CHANGES_REQUESTED** [review]
Interesting challenge with a good solution, minor changes requested. Thanks!

**Review by asagulenko — APPROVED** [review]

**Review by Ismatova-Rumiya — COMMENTED** [review]
Generally, after your previous trials,  you have great achievement in term of drafting your code, but minor change are required.

**Review by Ismatova-Rumiya — APPROVED** [review]

**Comment by Mushtary** [comment]
Thank you, @asagulenko and @Ismatova-Rumiya, for your review. I truly appreciate it!
